File: docs/lib/utils/catalogue_connector_utils.js
import { CatalogueConnector } from "../abstractions/catalogue_connector.js";
import { Result } from "../result.js";

class CatalogueConnectorUtils {

    /**
     * @template TItem
     * @template TFilters
     * @param {CatalogueConnector<TItem, TFilters>} catalogueConnector 
     * @param {string} id 
     * @param {TItem?} item 
     */
    static async syncItem(catalogueConnector, id, item) {
        let original = (await catalogueConnector.getAsync(id)).data;
        if (original == null) {
            if (item == null) return new Result({ data: null });
            else return await catalogueConnector.insertAsync(item);
        } else {
            if (item == null) return await catalogueConnector.deleteAsync(id);
            else if (item.modifiedAt < original.modifiedAt) return new Result({ data: original });
            else return await catalogueConnector.updateAsync(id, item);
        }
    }

}

export {
    CatalogueConnectorUtils
}

File: Core/Utils/CatalogueConnectorUtils.cs
using KolibSoft.Catalogue.Core.Abstractions;

namespace KolibSoft.Catalogue.Core.Utils;

public static class CatalogueConnectorUtils
{

    public static async Task<Result<TItem?>> SyncItem<TItem, TFilters>(this ICatalogueConnector<TItem, TFilters> catalogueConnector, Guid id, TItem? item)
        where TItem : IItem
    {
        var original = (await catalogueConnector.GetAsync(id)).Data;
        if (original == null)
        {
            if (item == null) return default;
            else return await catalogueConnector.InsertAsync(item);
        }
        else
        {
            if (item == null) return await catalogueConnector.DeleteAsync(id);
            else if (item.ModifiedAt < original.ModifiedAt) return original;
            else return await catalogueConnector.UpdateAsync(id, item);
        }
    }

}